### PR TITLE
`Adaptive learning`: Improve learning path graph view

### PR DESCRIPTION
**Pull request by MaximilianAnzinger**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This PR improves usability and delivers a more dynamic UI for the learning path graph view.

### Description

- nodes are now larger
- competency start nodes now display the competency progress cycles
- the view contains a mini-map to help the user navigate
- when clicking on competencies users get redirected to the corresponding competency card

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Course with LPs enabled, multiple competencies and relations

1. Log in to Artemis
2. Navigate to competency management
3. open the graph view of a student's learning path
4. zoom in and out and play with the mini-map, make sure it works as expected
5. make sure the corresponding competency rings are displayed for competency start nodes
6. Click on a competency node (start & end node) and check if you are redirected to the correct competency card

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

[Screencast](https://youtu.be/QkuQkIREoYk)

Screenshot:
chrome_KszM9wQY54

### CODE REVIEW

File: src/main/webapp/app/entities/competency.model.ts [hunk 1]
```diff
@@ -100,3 +100,16 @@ export function getIconTooltip(competencyTaxonomy?: CompetencyTaxonomy): string
...
+}
+
+export function getConfidence(competencyProgress: CompetencyProgress, masteryThreshold: number): number {
+    return Math.min(Math.round(((competencyProgress?.confidence ?? 0) / (masteryThreshold ?? 100)) * 100), 100);
+}
+
+export function getMastery(competencyProgress: CompetencyProgress, masteryThreshold: number): number {
+    const weight = 2 / 3;
+    return Math.round((1 - weight) * getProgress(competencyProgress) + weight * getConfidence(competencyProgress, masteryThreshold));
+}
```
Comment (JohannesStoehr) [inline src/main/webapp/app/entities/competency.model.ts:115]: Finally extracted into one function 🥳

**Review by JohannesStoehr — APPROVED** [review]
Code looks good to me

File: src/main/java/de/tum/in/www1/artemis/web/rest/LearningPathResource.java [hunk 2]
```diff
@@ -241,6 +244,33 @@ public ResponseEntity<Long> getLearningPathId(@PathVariable Long courseId) {
...
+            // update progress and construct DTOs
+            progressDTOs = learningPath.getCompetencies().stream().map(competency -> {
+                var progress = competencyProgressService.updateCompetencyProgress(competency.getId(), learningPath.getUser());
+                return new CompetencyProgressForLearningPathDTO(competency.getId(), competency.getMasteryThreshold(), progress.getProgress(), progress.getConfidence());
+            }).collect(Collectors.toSet());
+        }
+        else {
+            throw new AccessForbiddenException("You are not authorized to access other students competency progress.");
+        }
+        return ResponseEntity.ok(progressDTOs);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LearningPathResource.java:271]: style
```suggestion
        if (!(user.getId().equals(learningPath.getUser().getId()) || authorizationCheckService.isAtLeastInstructorInCourse(learningPath.getCourse(), user))) {
            throw new AccessForbiddenException("You are not authorized to access other students competency progress.");
        }
         // update progress and construct DTOs
        Set<CompetencyProgressForLearningPathDTO> progressDTOs = learningPath.getCompetencies().stream().map(competency -> {
            var progress = competencyProgressService.updateCompetencyProgress(competency.getId(), learningPath.getUser());
            return new CompetencyProgressForLearningPathDTO(competency.getId(), competency.getMasteryThreshold(), progress.getProgress(), progress.getConfidence());
        }).collect(Collectors.toSet());
        return ResponseEntity.ok(progressDTOs);
```

**Review by Strohgelaender — DISMISSED** [review]
lgtm

**Review by JohannesStoehr — DISMISSED** [review]

File: src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path-legend.component.html [hunk 3]
```diff
@@ -1,12 +1,4 @@
 <div class="container legend-wrapper">
-    <div class="row" ngbTooltip="{{ 'artemisApp.learningPath.graph.legend.competencyStart.tooltip' | artemisTranslate }}">
```
Comment (rstief) [inline src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path-legend.component.html:2]: Just a question: Why were competency starts completely removed? I don't quite understand that.

File: src/test/java/de/tum/in/www1/artemis/lecture/LearningPathIntegrationTest.java [hunk 4]
```diff
@@ -489,4 +494,33 @@ void getLearningPathIdNotExisting() throws Exception {
...
+    @Test
+    @WithMockUser(username = INSTRUCTOR_OF_COURSE, roles = "INSTRUCTOR")
+    void testGetCompetencyProgressForLearningPathByInstructor() throws Exception {
+        course = learningPathUtilService.enableAndGenerateLearningPathsForCourse(course);
+        final var student = userRepository.findOneByLogin(STUDENT_OF_COURSE).orElseThrow();
+        final var learningPath = learningPathRepository.findByCourseIdAndUserIdElseThrow(course.getId(), student.getId());
+        final var result = request.get("/api/learning-path/" + learningPath.getId() + "/competency-progress", HttpStatus.OK, Set.class);
+        assertThat(result).hasSize(5);
+    }
+
```
Comment (rstief) [inline src/test/java/de/tum/in/www1/artemis/lecture/LearningPathIntegrationTest.java:525]: The code for all three methods is pretty similar and could be extracted to another method. See the suggestion (if you add a return and the expected http code as aparmeter you could also add the case for the other student)

```suggestion
    @Test
    @WithMockUser(username = STUDENT_OF_COURSE, roles = "USER")
    void testGetCompetencyProgressForLearningPathByOwner() throws Exception {
        getCompetencyProgress();
    }

    @Test
    @WithMockUser(username = TEST_PREFIX + "student2", roles = "USER")
    void testGetCompetencyProgressForLearningPathByOtherStudent() throws Exception {
        course = learningPathUtilService.enableAndGenerateLearningPathsForCourse(course);
        final var student = userRepository.findOneByLogin(STUDENT_OF_COURSE).orElseThrow();
        final var learningPath = learningPathRepository.findByCourseIdAndUserIdElseThrow(course.getId(), student.getId());
        request.get("/api/learning-path/" + learningPath.getId() + "/competency-progress", HttpStatus.FORBIDDEN, Set.class);
    }

    @Test
    @WithMockUser(username = INSTRUCTOR_OF_COURSE, roles = "INSTRUCTOR")
    void testGetCompetencyProgressForLearningPathByInstructor() throws Exception {
        getCompetencyProgress();
    }

  void getCompetencyProgress() throws Exception {
    course = learningPathUtilService.enableAndGenerateLearningPathsForCourse(course);
        final var student = userRepository.findOneByLogin(STUDENT_OF_COURSE).orElseThrow();
        final var learningPath = learningPathRepository.findByCourseIdAndUserIdElseThrow(course.getId(), student.getId());
        assertThat(result).hasSize(5);
  }
```

**Review by rstief — CHANGES_REQUESTED** [review]
Looks good except for one small question and a small change 👍

File: src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path-legend.component.html (hunk 3 above)
Comment (MaximilianAnzinger) [reply src/main/webapp/app/course/learning-paths/learning-path-graph/learning-path-legend.component.html:2]: They're replaced by the competency rings.

**Review by rstief — APPROVED** [review]
Code LGTM 👍

**Review by Kroko-fant — DISMISSED** [review]
Tested in a testing session worked as described

**Review by basak-akan — APPROVED** [review]
Tested in testing session, really nice and dynamic UI for the learning path graph view!

**Review by rstief — APPROVED** [review]
Also tested manually in testing session.

**Review by lennart-keller — APPROVED** [review]
Tested in testing session. Works fine.

**Review by lennart-keller — APPROVED** [review]
Reapprove after base branch change

**Review by JohannesStoehr — APPROVED** [review]
Reapprove with new base branch

**Review by rstief — APPROVED** [review]
reapproved after base branch change

**Review by basak-akan — APPROVED** [review]
Reapproved after base branch change, retested on TS1

**Review by Strohgelaender — APPROVED** [review]

**Review by Kroko-fant — APPROVED** [review]
Reapprove after base branch change

**Pull request merged**